prov/shm/src/smr_ep.c

```diff
@@ -296,90 +343,193 @@ static void smr_progress_iov(struct smr_cmd *cmd, struct smr_ep_entry *entry,
+
+	for (iov_count = 0; iov_count < cmd->hdr.op.iov_count; iov_count++) {
+		ret = ofi_mr_verify(&domain->util_domain.mr_map,
+				cmd->data.rma_iov[iov_count].len,
+				(uintptr_t *) &(cmd->data.rma_iov[iov_count].addr),
+				cmd->data.rma_iov[iov_count].key,
+				cmd->hdr.op.op == ofi_op_write ?
+				FI_REMOTE_WRITE : FI_REMOTE_READ);
+		if (ret)
 			break;
 
-		match_attr.addr = cmd->hdr.op.addr;
-		match_attr.tag = cmd->hdr.op.tag;
+		iov[iov_count].iov_base = (void *) cmd->data.rma_iov[iov_count].addr;
+		iov[iov_count].iov_len = cmd->data.rma_iov[iov_count].len;
+	}
 
-		dlist_entry = dlist_remove_first_match(&recv_queue->recv_list,
-						       recv_queue->match_recv,
-						       &match_attr);
-		if (!dlist_entry) {
-			if (freestack_isempty(ep->unexp_fs))
-				break;
-			unexp = freestack_pop(ep->unexp_fs);
-			memcpy(&unexp->cmd, cmd, sizeof(*cmd));
-			ofi_cirque_discard(smr_cmd_queue(ep->region));
-			dlist_insert_tail(&unexp->entry, &ep->unexp_queue.msg_list);
-			continue;
-		}
-		entry = container_of(dlist_entry, struct smr_ep_entry, entry);
+	ofi_cirque_discard(smr_cmd_queue(ep->region));
+	cmd = ofi_cirque_head(smr_cmd_queue(ep->region));
+	if (cmd->hdr.op.op != ofi_op_write &&
+	    cmd->hdr.op.op != ofi_op_read_req)
+		return -FI_EINVAL;
+
+	switch (cmd->hdr.op.op_src) {
+	case smr_src_inline:
+		ret = smr_progress_inline(cmd, iov, iov_count, &total_len, ret);
+		break;
+	case smr_src_inject:
+		ret = smr_progress_inject(cmd, iov, iov_count, &total_len, ep, ret);
+		break;
+	case smr_src_iov:
+		ret = smr_progress_iov(cmd, iov, iov_count, &total_len, ep, ret);
+		break;
+	default:
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"unidentified operation type\n");
+		ret = -FI_EINVAL;
+	}
 
-		switch (cmd->hdr.op.op_src){
-		case smr_src_inline:
-			smr_progress_inline(cmd, entry);
+	return ret;
+}
+
+void smr_progress_cmd(struct smr_ep *ep)
+{
+	struct smr_cmd *cmd;
+	int ret = 0;
+
+	fastlock_acquire(&ep->region->lock);
+	fastlock_acquire(&ep->util_ep.rx_cq->cq_lock);
+
+	while (!ofi_cirque_isempty(smr_cmd_queue(ep->region))) {
+		cmd = ofi_cirque_head(smr_cmd_queue(ep->region));
+
+		switch (cmd->hdr.op.op) {
+		case ofi_op_msg:
+		case ofi_op_tagged:
+			ret = smr_progress_cmd_msg(ep, cmd);
 			break;
-		case smr_src_inject:
-			smr_progress_inject(cmd, entry, ep);
+		case ofi_op_write:
+		case ofi_op_read_req:
+			ret = smr_progress_cmd_rma(ep, cmd);
 			break;
-		case smr_src_iov:
-			smr_progress_iov(cmd, entry, ep);
+		case ofi_op_write_rsp:
+		case ofi_op_read_rsp:
 			break;
 		default:
 			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
 				"unidentified operation type\n");
-			entry->err = FI_EINVAL;
-		}
-		ret = ep->rx_comp(ep, entry->context, entry->flags,
-				  ret, entry->iov[0].iov_base, &addr, cmd->hdr.op.tag,
-				  entry->err);
-		if (ret) {
-			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
-				"unable to process rx completion\n");
+			ret = -FI_EINVAL;
 		}
-		freestack_push(ep->recv_fs, entry);
+		if (ret == -FI_ENOMSG)
```

Comment:
The switch statement above calls smr_progress_cmd_msg or *_rma.  It looks like ret can only equal FI_ENOMSG if we call smr_progress_cmd_msg().  If that's the case, move this check inside the case ofi_op_msg/tagged.

FI_ENOMSG doesn't make sense to me as a return code for progressing an RMA operation.  That's what made me notice this check.